website/addons/mendeley/views.py

```diff
@@ -0,0 +1,118 @@
+import httplib as http
+
+from flask import request
+
+from framework.exceptions import HTTPError
+from website.oauth.models import ExternalAccount
+from website.project.decorators import must_be_contributor_or_public
+from website.project.decorators import must_have_permission
+from website.project.decorators import must_not_be_registration
+from website.project.decorators import must_have_addon
+
+from .model import Mendeley
+
+
+@must_have_addon('mendeley', 'user')
+def list_mendeley_accounts_user(auth, user_addon):
+    return {
+        'accounts': [
+            {
+                'id': account._id,
+                'provider_id': account.provider_id,
+                'display_name': account.display_name,
+            } for account in auth.user.external_accounts
+            if account.provider == 'mendeley'
+        ]
+    }
+
+@must_have_permission('write')
+@must_have_addon('mendeley', 'node')
+@must_not_be_registration
+def list_mendeley_accounts_node(pid, auth, node, project, node_addon):
+    accounts = [
+        each for each in auth.user.external_accounts if each.provider == 'mendeley'
+    ]
+    if (
+        node_addon.external_account and
+        node_addon.external_account not in accounts
+    ):
+        accounts.append(node_addon.external_account)
+
+    return {
+        'accounts': [
+            {
+                'id': each._id,
+                'provider_id': each.provider_id,
+                'display_name': each.display_name,
+            } for each in accounts
+        ]
+    }
+
+@must_have_permission('write')
+@must_have_addon('mendeley', 'node')
+@must_not_be_registration
+def list_citationlists_node(pid, account_id, auth, node, project, node_addon):
+    # TODO: clean up signature
+
+    account = ExternalAccount.load(account_id)
```

Comment:
I'm wondering if this is completely secure. If I have the primary key of your Mendeley account, I can make a new project, add the Mendeley addon, and spoof a request to this view function with your primary key. Then I can see your folders. This could actually happen if you add me as a contributor on one of your projects, since I can grab the primary key of your Mendeley account from my network and store it for later attacks. I think it would be ideal if we never exposed the primary key of that record to anyone but its owner.

I would also think that only the user who provided the Mendeley credentials should be able to see the folders associated with the account. How about asserting that the Mendeley account targeted in this view belongs to the current user?